util/network-devp2p/src/ip_utils.rs

```diff
@@ -358,6 +365,86 @@ pub fn map_external_address(local: &NodeEndpoint) -> Option<NodeEndpoint> {
 	None
 }
 
+fn search_natpmp(local: &NodeEndpoint) -> Option<NodeEndpoint> {
+	if let SocketAddr::V4(ref local_addr) = local.address {
+		let local_port = local_addr.port();
+		let local_udp_port = local.udp_port;
+
+		let search_gateway_child = ::std::thread::spawn(move || {
+			let get_public_addr = |n: &mut Natpmp| {
+				// this function call want to receive `Response::Gateway` response from router, if other then it is an Error.
+				n.send_public_address_request()?;
+				::std::thread::sleep(Duration::from_millis(NAT_PMP_PORT_MAPPING_WAITING_DURATION));
+				match n.read_response_or_retry() {
+					Ok(Response::Gateway(gw)) => Ok(gw),
+					Err(e) => {
+						debug!(target: "network", "IP request error: {}", e);
+						Err(e)
+					},
+					_ => Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+				}
+			};
+
+			let mut n = Natpmp::new()?;
+
+			let gw = get_public_addr(&mut n)?;
```

Comment:
You could replace the function closure with a direct call as you did bellow IMO

Reply:
done